bin/node/runtime/src/lib.rs

```diff
@@ -536,28 +537,34 @@ parameter_types! {
 impl SignedExtensionProvider for Runtime {
 	type Extra = SignedExtra;
 
-	fn construct_extras(index: IndexFor<Self>) -> Self::Extra {
-		// take the biggest period possible.
-		let period = BlockHashCount::get()
-			.checked_next_power_of_two()
-			.map(|c| c / 2)
-			.unwrap_or(2) as u64;
-		let current_block = System::block_number()
-			.saturated_into::<u64>()
-			// The `System::block_number` is initialized with `n+1`,
-			// so the actual block number is `n`.
-			.saturating_sub(1);
+	fn construct_extras(index: IndexFor<Self>, era: Era, prior_block_hash: Option<Hash>) -> (
+		Self::Extra,
+		Option<<Self::Extra as SignedExtension>::AdditionalSigned>
+	) {
```

Comment:
Reading through the list of parameters and the way this is used I am wondering if this is the best way to structure the needed parameters. Just consider that any new chain might add further information or – as we do here – there is another parameter `tip`, that isn't actually covered.

I wonder if having the input be an associated type as well and constraint the setting of specific values via a trait might be more flexible and future proof. [Something like this maybe](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c4bfb69402cc1fb49405e13666073a6c)?

Reply:
hmm maybe i'm not seeing it, but why would that be preferable over say just passing a struct `ConstructExtrasParams` (which I remember thinking to do, but it skipped my mind) to the `construct_extras` method, which contains all the things it needs to actually `construct_extras` :D.

Reply:
I think we mean the same thing. I just wonder how to instantiate that other params-thing and still have it extensible. My current though it is, we might want the return of `construct_extras` to be a result, so that if thing are missing, we can fail appropriately.

Reply:
oh I was thinking more like:

```rust
#[derive(Default)]
pub struct ConstructExtrasParams<T: frame_system::Trait> {
    pub index: Option<IndexFor<T>>,
    pub era: Option<Era>,
    pub tip: Option<u64>,
    pub prior_block_hash: Option<Hash>,
    // so we can extend later on, 
    __nonexhaustive: ()
}

```